Src/HALAL/Services/Time/Scheduler.cpp

```diff
@@ -0,0 +1,291 @@
+/*
+ * Scheduler.hpp
+ *
+ * Created on: 17 nov. 2025
+ *     Author: Victor (coauthor Stephan)
+ */
+#include "HALAL/Services/Time/Scheduler.hpp"
+
+#ifndef TESTING_ENV
+    // This is needed to register a TimerPeripheral
+    #include "HALAL/Models/TimerPeripheral/TimerPeripheral.hpp"
+#endif
+#include <algorithm>
+#include <limits>
+
+
+/* NOTE(vic): Pido perdón a Boris pero es la mejor manera que se me ha ocurrido hacer esto */
+#define SCHEDULER_RCC_TIMER_ENABLE \
+    glue(glue(RCC_APB1LENR_TIM, SCHEDULER_TIMER_IDX), EN)
+#define SCHEDULER_GLOBAL_TIMER_IRQn \
+    glue(TIM, glue(SCHEDULER_TIMER_IDX, _IRQn))
+#define SCHEDULER_GLOBAL_TIMER_CALLBACK() \
+    extern "C" void glue(TIM, glue(SCHEDULER_TIMER_IDX, _IRQHandler))(void)
+
+#define Scheduler_global_timer ((TIM_TypeDef*)SCHEDULER_TIMER_BASE)
+namespace {
+constexpr uint64_t kMaxIntervalUs =
+    static_cast<uint64_t>(std::numeric_limits<uint32_t>::max()) + 1ULL;
+}
+
+std::array<Scheduler::Task, Scheduler::kMaxTasks> Scheduler::tasks_{};
+uint64_t Scheduler::sorted_task_ids_ = 0;
+std::size_t Scheduler::active_task_count_{0};
+
+uint32_t Scheduler::ready_bitmap_{0};
+uint32_t Scheduler::free_bitmap_{0xFFFF'FFFF};
+uint64_t Scheduler::global_tick_us_{0};
+uint64_t Scheduler::current_interval_us_{0};
+
+inline uint8_t Scheduler::get_at(uint8_t idx) {
+    int word_idx = idx > 7;
+    uint32_t shift = (idx & 7) << 2;
+    return (((uint32_t*)sorted_task_ids_)[word_idx] & (0x0F << shift)) >> shift;
+}
+inline void Scheduler::set_at(uint8_t idx, uint8_t id) {
+    uint32_t shift = idx*4;
+    uint64_t clearmask = ~(0xFF << shift);
+    Scheduler::sorted_task_ids_ = (sorted_task_ids_ & clearmask) | (id << shift);
+    // sorted_task_ids_ |= ((id & 0x0F) << shift); // This is also an option in case id is incorrect, I don't think it's necessary though
+}
+inline uint8_t Scheduler::front_id() {
+    return ((uint32_t*)sorted_task_ids_)[0] & 0xF;
+}
+inline void Scheduler::pop_front() {
+    // O(1) remove of logical index 0
+    Scheduler::active_task_count_--;

```

Comment:
There is no point in substracting this here,

There is two functions, remove from sorted list and deallocate,

Deallocate frees a slot in the task array

Remove from sorted list does not free a slot from the task array, only removes a task from the sorted list. This function is a fastpath of remove sorted, you have to be consistent with this counter:
You either increment and decrement it only in the allocate-deallocate functions, or only in the insert-remove functions, both are fine but now this is inconsistent